contrib/envoyproxy/envoy/fakehttp.go

```diff
@@ -0,0 +1,189 @@
+	extproc "github.com/envoyproxy/go-control-plane/envoy/service/ext_proc/v3"
+	"google.golang.org/grpc/metadata"
+)
+
+// checkPseudoRequestHeaders Verify the required HTTP2 headers are present
+// Some mandatory headers need to be set. It can happen when it wasn't a real HTTP2 request sent by Envoy,
+func checkPseudoRequestHeaders(headers map[string]string) error {
+	for _, header := range []string{":authority", ":scheme", ":path", ":method"} {
+		if _, ok := headers[header]; !ok {
+			return fmt.Errorf("missing required headers: %q", header)
+		}
+	}
+
+	return nil
+}
+
+// checkPseudoResponseHeaders Verify the required HTTP2 headers are present
+// Some mandatory headers need to be set. It can happen when it wasn't a real HTTP2 request sent by Envoy,
+func checkPseudoResponseHeaders(headers map[string]string) error {
+	if _, ok := headers[":status"]; !ok {
+		return fmt.Errorf("missing required ':status' headers")
+	}
+
+	return nil
+}
+
+func getRemoteAddr(md metadata.MD) string {
+	xfwd := md.Get("x-forwarded-for")
+	length := len(xfwd)
+	if length == 0 {
+		return ""
+	}
+
+	// Get the first right value of x-forwarded-for headers
+	// The rightmost IP address is the one that will be used as the remote client IP
+	// https://datadoghq.atlassian.net/wiki/spaces/TS/pages/2766733526/Sensitive+IP+information#Where-does-the-value-of-the-http.client_ip-tag-come-from%3F
+	return xfwd[length-1]
+}
+
+// partitionPeusdoHeaders Separate normal headers of the initial request made by the client and the pseudo headers of HTTP/2
+// - Format the headers to be used by the tracer as a map[string][]string
+// - Set headers keys to be canonical
+func partitionPeusdoHeaders(receivedHeaders []*corev3.HeaderValue) (map[string][]string, map[string]string) {
+	headers := make(map[string][]string, len(receivedHeaders)-4)
+	pseudoHeaders := make(map[string]string, 4)
+	for _, v := range receivedHeaders {
+		key := v.GetKey()
+		if key == "" {
+			continue
+		}
+		if key[0] == ':' {
+			pseudoHeaders[key] = string(v.GetRawValue())
+			continue
+		}
+
+		headers[http.CanonicalHeaderKey(key)] = []string{string(v.GetRawValue())}
+	}
+	return headers, pseudoHeaders
+}
+
+func NewFakeResponseWriterFromExtProc(w http.ResponseWriter, res *extproc.ProcessingRequest_ResponseHeaders) error {
+	headers, pseudoHeaders := partitionPeusdoHeaders(res.ResponseHeaders.GetHeaders().GetHeaders())
+
+	if err := checkPseudoResponseHeaders(pseudoHeaders); err != nil {
+		return err
+	}
+
+	status, err := strconv.Atoi(pseudoHeaders[":status"])
+	if err != nil {
+		return fmt.Errorf("error parsing status code %q: %w", pseudoHeaders[":status"], err)
+	}
+
+	for k, v := range headers {
+		w.Header().Set(k, strings.Join(v, ","))
+	}
+
+	w.WriteHeader(status)
+	return nil
+}
+
+// NewRequestFromExtProc creates a new http.Request from an ext_proc RequestHeaders message
+func NewRequestFromExtProc(ctx context.Context, req *extproc.ProcessingRequest_RequestHeaders) (*http.Request, error) {
```

Comment:
```suggestion
func newRequest(ctx context.Context, req *extproc.ProcessingRequest_RequestHeaders) (*http.Request, error) {
```

Reply:
Applied in db73a1f13bf8855a9c1a351a685f38629d5926bc thanks 😄